File: Pharmaceutical_Company_Example/Behavioral/Python/Iterator.py
from typing import List, Dict, Optional

# Iterator Interface
class ApplicationIterator:
    def has_next(self) -> bool:
        raise NotImplementedError("This method should be overridden!")

    def is_next(self) -> Optional[Dict[str, str]]:
        raise NotImplementedError("This method should be overridden!")

# Concrete Iterator
class DrugApplicationIterator(ApplicationIterator):
    def __init__(self, applications: List[Dict[str, str]]):
        self.applications = applications
        self.position = 0

    def has_next(self) -> bool:
        return self.position < len(self.applications)

    def is_next(self) -> Optional[Dict[str, str]]:
        if self.has_next():
            app = self.applications[self.position]
            self.position += 1
            return app
        return None

# Aggregate Interface
class ApplicationCollection:
    def create_iterator(self) -> ApplicationIterator:
        raise NotImplementedError("This method should be overridden!")

# Concrete Aggregate
class DrugApplicationCollection(ApplicationCollection):
    def __init__(self):
        self.applications = []

    def add_application(self, application: Dict[str, str]):
        self.applications.append(application)

    def create_iterator(self) -> DrugApplicationIterator:
        return DrugApplicationIterator(self.applications)


drug_collection = DrugApplicationCollection()

drug_collection.add_application({"drug_name": "Test Vaccine 01", "status": "Under Review"})
drug_collection.add_application({"drug_name": "Test Small Molecule 01", "status": "Approved"})
drug_collection.add_application({"drug_name": "Test Monoclonal Antibody 01", "status": "Rejected"})

iterator = drug_collection.create_iterator()

# Use the iterator to traverse the applications
while iterator.has_next():
    application = iterator.is_next()
    if application:
        print(f"Drug Name: {application['drug_name']} - Status: {application['status']}")

File: Pharmaceutical_Company_Example/Behavioral/R/Iterator.R
# Iterator Interface
ApplicationIterator <- setRefClass(
  "ApplicationIterator",
  methods = list(
    has_next = function() {
      stop("This method should be overridden!")
    },
    is_next = function() {
      stop("This method should be overridden!")
    }
  )
)

# Concrete Iterator
DrugApplicationIterator <- setRefClass(
  "DrugApplicationIterator",
  contains = "ApplicationIterator",
  fields = list(
    applications = "list",
    position = "numeric"
  ),
  methods = list(
    initialize = function(applications) {
      .self$applications <- applications
      .self$position <- 1
    },
    has_next = function() {
      return(position <= length(applications))
    },
    is_next = function() {
      if (has_next()) {
        app <- applications[[position]]
        position <<- position + 1
        return(app)
      } else {
        return(NULL)
      }
    }
  )
)

# Aggregate Interface
ApplicationCollection <- setRefClass(
  "ApplicationCollection",
  methods = list(
    create_iterator = function() {
      stop("This method should be overridden!")
    }
  )
)

# Concrete Aggregate
DrugApplicationCollection <- setRefClass(
  "DrugApplicationCollection",
  contains = "ApplicationCollection",
  fields = list(applications = "list"),
  methods = list(
    add_application = function(application) {
      applications[[length(applications) + 1]] <<- application
    },
    create_iterator = function() {
      return(DrugApplicationIterator$new(applications = applications))
    }
  )
)

drug_collection <- DrugApplicationCollection$new()

drug_collection$add_application(list(drug_name = "Test Vaccine 01", status = "Under Review"))
drug_collection$add_application(list(drug_name = "Test Small Molecule 01", status = "Approved"))
drug_collection$add_application(list(drug_name = "Test Monoclonal Antibody 01", status = "Rejected"))

iterator <- drug_collection$create_iterator()

# Use the iterator to traverse the applications
while (iterator$has_next()) {
  application <- iterator$is_next()
  message("Drug Name:", application$drug_name, "- Status:", application$status)
}
